driver/generic/st77xx.py

```diff
@@ -0,0 +1,404 @@
+        self.buf4 = bytearray(4)
+
+        self.cs,self.dc,self.bl,self.rst=[(machine.Pin(p,machine.Pin.OUT) if isinstance(p,int) else p) for p in (cs,dc,bl,rst)]
+        self.bl.value(1)
+        self.bl=machine.PWM(self.bl)
+        self.rot=rot
+        self.width,self.height=(0,0) # set in apply_rotation
+
+        if res not in suppRes: raise ValueError('Unsupported resolution %s; the driver supports: %s.'%(str(res),', '.join(str(r) for r in suppRes)))
+        if suppModel and model not in suppModel: raise ValueError('Unsupported model %s; the driver supports: %s.'%(str(model),', '.join(str(r) for r in suppModel)))
+
+        self.res=res
+        self.model=model
+
+        self.dma=dma
+        self.spi=spi
+        self.hard_reset()
+
+    def off(self):
+        self.bl.value(0)
+    def hard_reset(self):
+        if self.rst:
+            for v in (1,0,1):
+                self.rst.value(v)
+                time.sleep(.2)
+            time.sleep(.2)
+        self.config()
+    def set_backlight(self,percent):
+        self.bl.duty_u16(percent*655)
+    def set_window(self, x, y, w, h):
+        c0,r0=ST77XX_COL_ROW_MODEL_START_ROTMAP[self.res[0],self.res[1],self.model][self.rot%4]
+        struct.pack_into('>hh', self.buf4, 0, c0+x, c0+x+w-1)
+        self.write_register(ST77XX_CASET, self.buf4)
+        struct.pack_into('>hh', self.buf4, 0, r0+y, r0+y+h-1)
+        self.write_register(ST77XX_RASET, self.buf4)
+    def apply_rotation(self,rot):
+        self.rot=rot
+        if (self.rot%2)==0: self.width,self.height=self.res
+        else: self.height,self.width=self.res
+        self.write_register(ST77XX_MADCTL,bytes([ST77XX_MADCTL_BGR | ST77XX_MADCTL_ROTS[self.rot%4]]))
+
+    def blit(self, x, y, w, h, buf, is_blocking=True):
+        self.set_window(x, y, w, h)
+        if self.dma: self.write_register_dma(ST77XX_RAMWR, buf, is_blocking)
+        else: self.write_register(ST77XX_RAMWR, buf)
+
+    def clear(self, color):
+        bs=128 # write pixels in chunks; makes the fill much faster
+        struct.pack_into('>h',self.buf2,0,color)
+        buf=bs*bytes(self.buf2)
+        npx=self.width*self.height
+        self.set_window(0, 0, self.width, self.height)
+        self.write_register(ST77XX_RAMWR, None)
+        self.cs.value(0)
+        self.dc.value(1)
+        for _ in range(npx//bs): self.spi.write(buf)
+        for _ in range(npx%bs): self.spi.write(self.buf2)
+        self.cs.value(1)
+
+    def write_register(self, reg, buf=None):
+        struct.pack_into('B', self.buf1, 0, reg)
+        self.cs.value(0)
+        self.dc.value(0)
+        self.spi.write(self.buf1)
+        if buf is not None:
+            self.dc.value(1)
+            self.spi.write(buf)
+        self.cs.value(1)
+
+    def write_register_dma(self, reg, buf, is_blocking=True):
+        'If *is_blocking* is False, used should call wait_dma explicitly.'
+        SPI1_BASE = 0x40040000 # FIXME: will be different for another SPI bus?
```

Comment:
It is port-specific (I might rename the function to `rp2_write_register_dma`) but it will only be called if the `dma` object (from `rp2_dma`) is provided in the ctor. I will isolate that piece somewhat more visibly, but I'd keep it in the generic file, as it is just a few lines.